### ISSUE TITLE
Layout/ClassStructure doesn't work when constants use the initializer

### ISSUE THREAD

**Issue by tjschuck** [rubocop/rubocop#11288]
Not sure if this is even fixable or not!

Right now, Layout/ClassStructure wants all `constants` higher in the file than the `initializer`.  However, sometimes constants _use_ the initializer, and if they're defined above it, they don't use it and throw an `ArgumentError: wrong number of arguments (given X, expected 0)` because they're still trying to use the default object initializer.

Consider:

```ruby
class PageSize
  def initialize(width, height)
    @width = width
    @height = height
  end

  A4 = new(8.26, 11.69)
  LETTER = new(8.50, 11.00)

  def ratio
    @width / @height
  end
end
```

Rubocop fails this with:

```
app/models/page_size.rb:7:3: C: Layout/ClassStructure: constants is supposed to appear before initializer.
  A4 = new(8.26, 11.69)
  ^^^^^^^^^^^^^^^^^^^^^
```

Moving the constants before the initializer passes Layout/ClassStructure, but then `PageSize::A4` throws `ArgumentError: wrong number of arguments (given 2, expected 0)`

--------

## Expected behavior

Honestly, I don't know!  Let constants-with-`new` be below the initializer?

## Actual behavior

See above.

## Steps to reproduce the problem

See above.

## RuboCop version

```
1.40.0 (using Parser 3.1.3.0, rubocop-ast 1.24.0, running on ruby 3.0.4) [x86_64-darwin21]
```

**Comment by vlad-pisanov** [comment]
that seems like an edge case; I'd just disable the cop for that one line
```ruby
A4 = new(8.26, 11.69) # rubocop:disable Layout/ClassStructure
```

**Comment by pboling** [comment]
Agree.  The `Layout/ClassStructure` defined as "GOOD" should remain as is, and edge cases should be designated as such by disabling the cop locally.

**Comment by pboling** [comment]
I think the rule hints at a potential better design here, that would result in the linter not failing.
```
class Page
  class PageSize
    def initialize(width, height)
      @width = width
      @height = height
    end

    def ratio
      @width / @height
    end
  end

  A4 = PageSize.new(8.26, 11.69)
  LETTER = PageSize.new(8.50, 11.00)

  def initialize(size = A4)
    @size = size
  end
end
```

( I agree that this should be closed )
